### PR TITLE
Run part of `rustc_codegen_gcc`'s tests in CI

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Thanks to #112701 and @bjorn3, it made this much easier.

Also cc @antoyo.

r? @bjorn3

### CODE REVIEW

File: src/ci/run.sh
Comment (bjorn3) [inline src/ci/run.sh:127]: typo I made myself: in on

File: src/bootstrap/src/core/build_steps/test.rs
Comment (bjorn3) [inline src/bootstrap/src/core/build_steps/test.rs:1572]: rust-demangler is intentionally on stage 0 to avoid unnecessary recompilations: https://github.com/rust-lang/rust/pull/117197

File: src/bootstrap/src/core/build_steps/test.rs
Comment (bjorn3) [inline src/bootstrap/src/core/build_steps/test.rs:3002]: Did you mess up a rebase? I made the exact opposite change yesterday.

**Comment by bjorn3** [comment]
At least the x86_64-gnu-llvm-15 runner doesn't have libgccjit installed.

File: src/bootstrap/src/core/build_steps/test.rs
Comment (lqd) [inline src/bootstrap/src/core/build_steps/test.rs:632]: That also looks like a rebase problem.

File: src/bootstrap/src/core/build_steps/test.rs
Comment (lqd) [inline src/bootstrap/src/core/build_steps/test.rs:3203]: (Stray debugging code)

File: src/bootstrap/src/core/build_steps/test.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/test.rs:3002]: I completely did...

File: src/bootstrap/src/core/build_steps/test.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/test.rs:3203]: No it comes from https://github.com/rust-lang/rust/pull/112701/files#diff-8479eab02701e686aedb15b567dc8fc31220c6e4efb9565ccc9d662b7fee2214R3065. I keep it as close as possible to cranelift implementation.

File: src/ci/run.sh
Comment (GuillaumeGomez) [reply src/ci/run.sh:127]: Good catch haha.

File: src/bootstrap/src/core/build_steps/test.rs
Comment (lqd) [reply src/bootstrap/src/core/build_steps/test.rs:3212]: Then I guess one of you two could either:
- add a comment explaining why this commented code needs to be kept (and copied) to both places (maybe like "when vendoring is fixed replace X with the commented out code below" -- but actually correct about why it's been added and I don't know this)
- or else remove both copies of it?

Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/test.rs:3212]: I'm fine with both solution. A preference maybe @bjorn3 ?

Comment (bjorn3) [reply src/bootstrap/src/core/build_steps/test.rs:3212]: A comment would be something along the lines of:

> Uncomment the `prepare` command below once vendoring is implemented.

File: src/bootstrap/src/core/build_steps/test.rs
Comment (bjorn3) [inline src/bootstrap/src/core/build_steps/test.rs:3142]: You will want to adapt this for the targets cg_gcc actually supports, which may well be removing this all if cg_gcc supports all rustc targets.

Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/test.rs:3143]: @antoyo ^ ?

File: src/bootstrap/src/core/build_steps/test.rs
Comment (bjorn3) [inline src/bootstrap/src/core/build_steps/test.rs:3169]: You don't need to compile the standard library itself with `-Zpanic-abort-tests` and (for as long as llvm is set as default backend) `-Cpanic=abort`. `-Zpanic-abort-tests` is ignored when not building with `--test` to produce a test executable. As for the `-Csymbol-mangling-version=v0`, won't this overwrite the regular standard library copy?

File: src/bootstrap/src/core/build_steps/test.rs
Comment (bjorn3) [inline src/bootstrap/src/core/build_steps/test.rs:3194]: cg_gcc's build system doesn't need to be compiled with `-Cpanic=abort`, right?

**Comment by bjorn3** [comment]
```
2023-10-28T16:54:08.4668438Z   = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/arbitrary_self_types_pointers_and_wrappers.arbitrary_self_types_pointers_and_wrappers.4f746435a173466d-cgu.0.rcgu.o: in function `<i32 as core::fmt::Debug>::fmt':
2023-10-28T16:54:08.4669301Z           fake.c:(.text._RNvXs1g_NtNtCseVSX39tBz28_4core3fmt3numlNtB8_5Debug3fmtCs6OVZ8sMh4wz_42arbitrary_self_types_pointers_and_wrappers+0x5f): undefined reference to `core::fmt::num::__LT_impl_u20_core__fmt__LowerHex_u20_for_u20_i32_GT_::fmt'
2023-10-28T16:54:08.4670212Z           /usr/bin/ld: fake.c:(.text._RNvXs1g_NtNtCseVSX39tBz28_4core3fmt3numlNtB8_5Debug3fmtCs6OVZ8sMh4wz_42arbitrary_self_types_pointers_and_wrappers+0x81): undefined reference to `core::fmt::num::imp::__LT_impl_u20_core__fmt__Display_u20_for_u20_i32_GT_::fmt'
2023-10-28T16:54:08.4671109Z           /usr/bin/ld: fake.c:(.text._RNvXs1g_NtNtCseVSX39tBz28_4core3fmt3numlNtB8_5Debug3fmtCs6OVZ8sMh4wz_42arbitrary_self_types_pointers_and_wrappers+0xa3): undefined reference to `core::fmt::num::__LT_impl_u20_core__fmt__UpperHex_u20_for_u20_i32_GT_::fmt'
2023-10-28T16:54:08.4671270Z           collect2: error: ld returned 1 exit status
```

I presume it is trying to use a standard library compiled without `-Csymbol-mangling-version=v0`?

**Comment by GuillaumeGomez** [comment]
> I presume it is trying to use a standard library compiled without `-Csymbol-mangling-version=v0`?

Yes, but I don't understand how it's possible. The `RUSTC_SYSROOT` env variable is set and the `-Zcodegen-backend=gcc` flag is passed so I suppose it'll use this one. Maybe I'm missing something.

> By the way should cg_gcc maybe error out on `-Csymbol-mangling-version=legacy`?

I'll let @antoyo decide on that.

File: src/bootstrap/src/core/build_steps/test.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/test.rs:3169]: We need `-Cpanic=abort` because of how we (currently) handle it in cg_gcc. However I don't think `-Zpanic-abort-tests` is needed.

> As for the -Csymbol-mangling-version=v0, won't this overwrite the regular standard library copy?

Yes and it's what we need (currently). We need to fix that later on.

File: src/bootstrap/src/core/build_steps/test.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/test.rs:3194]: We do, otherwise we get:

```
error: the crate `panic_unwind` does not have the panic strategy `unwind`
```

File: src/bootstrap/src/core/build_steps/test.rs
Comment (antoyo) [reply src/bootstrap/src/core/build_steps/test.rs:3143]: Since we didn't build any cross-compilers, I assume we should only keep `x86_64`.

**Comment by antoyo** [comment]
> > By the way should cg_gcc maybe error out on `-Csymbol-mangling-version=legacy`?
> 
> I'll let @antoyo decide on that.

I don't mind keeping it available since it partly works.

**Comment by GuillaumeGomez** [comment]
It's working! \o/

Time for the bootstrap people to take a look now. :)

r? @onur-ozkan

**Comment by oli-obk** [comment]
How easy is it to run those tests locally? If I try to run them, but haven't installed the right packages, do I get a hint like with the Rustdoc tests?

**Comment by GuillaumeGomez** [comment]
The linker will fail saying that it cannot find `libgccjit`. Apart from that, I don't think we can do much more unfortunately... :-/ If you have ideas, please share. :)

File: src/bootstrap/src/core/build_steps/test.rs
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/test.rs:3066]: Would be nice to have `FIXME` header here

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-15/script.sh
Comment (onur-ozkan) [inline src/ci/docker/host-x86_64/x86_64-gnu-llvm-15/script.sh:7]: fyi: This script used for llvm16 and llvm17 too.

File: src/bootstrap/src/core/build_steps/test.rs
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/test.rs:3196]: Same as https://github.com/rust-lang/rust/pull/117313#discussion_r1378216479

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-15/script.sh
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-15/script.sh:7]: As discussed with @onur-ozkan, I'll set an environment variable to only run gcc tests in CI.

Comment (onur-ozkan) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-15/script.sh:7]: I think we should use `SKIP_CODEGEN_TESTS` condition here too

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-15/script.sh:7]: Yes that's the only place where I changed it in fact. ;)

**Comment by GuillaumeGomez** [comment]
Time to try with a broader check.

@bors try

**Comment by GuillaumeGomez** [comment]
Try passed but I forgot to add a `SKIP_CODEGEN_TESTS` check, so just in case.

@bors try

**Comment by GuillaumeGomez** [comment]
Seems like everything's working so far. Waiting to hear from @onur-ozkan.

**Review by onur-ozkan — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I forgot to add the condition when adding gcc backend in `run.sh`. Should be good now (let's wait for CI to confirm).

**Comment by GuillaumeGomez** [comment]
Got the approval from @onur-ozkan so let's try again with the fixed condition.

@bors r=onur-ozkan

**Comment by RalfJung** [comment]
How hard will it be for contributors to reproduce and fix these test failures if they show up in their PRs? Seems like libgccjit needs to be installed, how easy is that to do on Linux/macOS/Windows? Anything else that needs to be done, like changing `config.toml`?

We should be mindful of the potential of new contributor roadblocks from new CI checks that people can't easily debug locally.

**Comment by GuillaumeGomez** [comment]
You need to update your `config.toml` and add:

```toml
codegen-backends = ["gcc"]
```

Then you can run:

```
./x.py test compiler/rustc_codegen_gcc/ --stage 1
```

We only expect such failures to occur in case rustc's API changed, so just ensuring that it builds (and then fail when trying to link) is good enough.

However if you need something more advanced, you'll need to install the `libgccjit` package on your system and ensure that the above command worked.

The goal of running parts of cg_gcc in the rustc testsuite is to makes our lives easier when syncing with rustc because we currently have a lot of broken APIs. Going through such changes one by one take quite some time when we didn't sync in a long time.

And if it comes to the worse, they can always ask me or @antoyo to take a look.

**Comment by GuillaumeGomez** [comment]
> For that, it's sufficient to run `./x.py check`, it doesn't need the test suite, right? Don't we already check-build gcc on CI? (AFAIK we do check-build cranelift.)

The goal is to ensure that a change doesn't break things on our side too. We run some basic tests which is just a subset of all the tests we have in cg_gcc. The goal is to reduce sync burden for us. So in this regard, we followed what cg_cranelift did (I wrote this PR with @bjorn3's help).

> It would be good to find a way to make this information accessible to an unsuspecting contributor that receives a GCC backend CI failure in their PR. It seems reasonable to expect them to try `./x.py test compiler/rustc_codegen_gcc/`, but from there on our tooling should do its best to help them.

This is a very good point! Do you know where such information could be added? As for the tooling, next step for me will be to automate the download and "installation" (putting it into rustc's sysroot in short) of our libgccjit version. But for that I need a bit more work in cg_gcc first. But yes, I don't want anything manual involved for "outside users". Just running `x.py` should be enough.

> That's great. :) But also hard to remember. Would it makes sense to have @rust-lang/gcc-backend or so, something that the reviewer can find via auto-completion if they are out of their depth and want to page in more help? Or at least a rustbot [ping group](https://rustc-dev-guide.rust-lang.org/notification-groups/about.html)?

I'd like to have a rust backends team at some point considering we already have 4 (that I know of). But not sure if that makes sense or not. For now, having an alias for cg_gcc people would be quite nice. :)

**Comment by RalfJung** [comment]
> The goal is to ensure that a change doesn't break things on our side too.

Okay so there are common breakages that are not detected by `./x.py check`, but still generally easy to take care of by people that know nothing about the backend?

> Do you know where such information could be added? 

The error message that shows up when `./x.py test rustc_codegen_gcc` fails to find libgccjit, would be a good place IMO. And maybe somewhere in the dev guide? But I'm not sure if people will go looking there.

> we already have 4 backends (that I know of)

Then you are 1 ahead of me, I only know of 3. ;)

**Comment by GuillaumeGomez** [comment]
> Okay so there are common breakages that are not detected by `./x.py check`, but still generally easy to take care of by people that know nothing about the backend?

Sorry I'm a bit lost here. `x.py check` doesn't check that backends compile. Maybe there is a misunderstanding here (maybe on my side)?

The only case where a backend can stop compiling is if rustc changes an API without updating the backend. It happened quite often (not blaming everyone, if CI says everything is ok, then the problem is the CI :smile: ).


> The error message that shows up when `./x.py test rustc_codegen_gcc` fails to find libgccjit, would be a good place IMO. 

Not really, it's a linker error, so not under our control so to speak.

> And maybe somewhere in the dev guide? But I'm not sure if people will go looking there.

This is a very good idea!

> Then you are 1 ahead of me, I only know of 3. ;)

LLVM, cranelift and GCC for the 3 known ones. I was also counting miri, but maybe it doesn't match the backend definition.

**Comment by antoyo** [comment]
> Sorry I'm a bit lost here. `x.py check` doesn't check that backends compile. Maybe there is a misunderstanding here (maybe on my side)?

I don't think that's true.

The actual problem is that the API is changed and this breaks the codegen at run-time. So `check` doesn't help since it only checks that it compiles.

**Comment by GuillaumeGomez** [comment]
> I don't think that's true.
> 
> The actual problem is that the API is changed and this breaks the codegen at run-time. So `check` doesn't help since it only checks that it compiles.

Ok so the confusion was on my side. Thanks for clarifying!

**Comment by GuillaumeGomez** [comment]
> That also sounds quite hard to fix then without GCC knowledge, but it also should be quite rare.

If it can't be solved easily, contributors can come ask us. ;)

> `cmake`/`./configure` usually detect absence of a required library and can then show specific errors; `x.py` could do the same?

I'd really prefer to avoid that as much as possible. Finding if something is installed, whatever the OS sounds very much like a nightmare. ^^'

> Miri isn't quite a backend... it has a lot of things in common with the backends though, and ideally we'd be able to share more code, but for other things interpretation vs compilation does make a big difference.

We figured out why my count was wrong. :laughing:

**Comment by GuillaumeGomez** [comment]
So `llvm-15` is run differently in some other parts of the CI apparently. Well, more uncovering to do.

**Comment by antoyo** [comment]
> That also sounds quite hard to fix then without GCC knowledge, but it also should be quite rare.

It should be much easier if the scope is limited to a single PR than many months worth of commits (as happens for syncs).

**Comment by GuillaumeGomez** [comment]
Found out the issue: I was enforcing the mangling version, even when we were not running cg_gcc tests, which is wrong. I moved setting this value outside of the docker file.

**Comment by GuillaumeGomez** [comment]
cg_gcc tests passed again by modifying where we set the environment variables so time to give it another try.

@bors r=onur-ozkan

**Comment by GuillaumeGomez** [comment]
It finally passed! \o/

**Pull request merged**
